### ISSUE TITLE
Provenance Audit: Map metadata across the compilation pipeline

### ISSUE THREAD

**Issue by kaoutar55** [torch-spyre/torch-spyre#2574]
## Overview

Trace a reference model through every stage of the Spyre compilation pipeline and document where provenance metadata exists, where it transforms, and where it drops. This audit establishes the ground truth for all subsequent implementation work.

## Acceptance Criteria

- [x] Select a reference model (single linear layer or equivalent) and trace it through all six stages:
  1. PyTorch model source
  2. FX graph (pre-grad and post-grad)
  3. Inductor passes
  4. LoopLevelIR
  5. OpSpec
  6. SuperDSC JSON
- [x] At each stage, document:
  - Which provenance fields are present (e.g., `stack_trace`, `original_aten`, `from_node`, `origins`, `origin_node`, `traceback`)
  - The data structure and format of each field
  - How the field was populated (copied, derived, or generated)
  - Whether the field survives to the next stage
- [x] Identify the exact code locations where metadata is dropped or fails to propagate
- [x] Produce a pipeline diagram showing metadata flow (aligned with `docs/source/compiler/architecture.md`)
- [x] Produce an accompanying table: stage × metadata field → present / absent / transformed

## Key Files to Investigate

| Stage | Key files |
|-------|-----------|
| FX metadata | Upstream PyTorch Inductor (node.meta fields) |
| LoopLevelIR | `torch_spyre/_inductor/scratchpad.py` |
| Scheduler metadata | `torch_spyre/_inductor/scheduler.py` (`get_kernel_metadata()`) |
| OpSpec | `torch_spyre/_inductor/op_spec.py` |
| SuperDSC codegen | `torch_spyre/_inductor/codegen/superdsc.py`, `torch_spyre/_inductor/codegen/compute_ops.py` |

## Timeline

May 26 – June 12, 2026 (3 weeks)

## Notes

- Use `TORCH_COMPILE_DEBUG=1` and `TORCH_LOGS="+inductor"` to capture intermediate artifacts
- Compare against upstream Inductor's `inductor_provenance_tracking_node_mappings.json` for field naming conventions
- The audit should answer: "If I add a `debug_handle` at OpSpec, what upstream fields can I derive it from?"

**Comment by kaoutar55** [comment]
@gsmoon97 please add details about the findings you have for this task in the comment section.

**Comment by kaoutar55** [comment]
Hi @gsmoon97,

Hey Geonsik,

The provenance audit looks solid. The stage × field matrix confirmed what we suspected: `origins` is present 7/7 through the Inductor passes, while `origin_node` and `from_node` are fragile or run-dependent. That gives us a direct path into Phase 2.

For the next issue, you need to thread provenance data from the Inductor IR into the serialized `sdsc_*.json` so downstream tools can map a compiled kernel back to the user's source code. Three changes, done in sequence:

---

### 1) Add a `provenance` field to the OpSpec dataclass

`torch_spyre/_inductor/op_spec.py`, lines 51–73

A list of dicts works well. Each entry holds the `original_aten` op name and the `stack_trace` string when available:

```python
provenance: list[dict[str, str]] = field(default_factory=list)
```

---

### 2) Populate it in `create_op_spec`

`torch_spyre/_inductor/spyre_kernel.py`, lines 487–561

`self.current_node` is the `ComputedBuffer`. Call `.get_origins()` to get the FX node set, then pull metadata from each node:

```python
origins = self.current_node.get_origins()
provenance = []
for node in origins:
    entry = {}
    if hasattr(node, "meta"):
        if "original_aten" in node.meta:
            entry["original_aten"] = str(node.meta["original_aten"])
        if "stack_trace" in node.meta:
            entry["stack_trace"] = node.meta["stack_trace"]
    if entry:
        provenance.append(entry)
```

Then pass `provenance=provenance` to the `OpSpec(...)` constructor.

---

### 3) Serialize into sdsc_*.json

`torch_spyre/_inductor/codegen/compute_ops.py`, inside `generate_sdsc()`

Add a `"provenance_"` key to the per-op JSON object. A JSON list of `{original_aten, stack_trace}` entries, same structure as above. No new format needed.

---

### Validation

Use your SimpleMLP model. It compiles to 2 fused SuperDSC kernels, so you should see provenance entries for `aten.addmm` (or `aten.mm` + `aten.add` if decomposed) and `aten.relu` in the output JSON. Print `op_spec.provenance` right after construction to confirm data flows before you touch serialization.

---

### Watch out for the addmm decomposition

Your audit flagged that `stack_trace` is lost on mm/add when addmm decomposes. The `origins` set still has the nodes, but their `stack_trace` field can be empty. Just skip the key or set it to `null` in that case. A missing stack trace should not block the rest of the provenance entry.

---

### Suggested order

Do steps 1 and 2 together, verify with a print statement, then move to step 3. That way you know the data is correct before you deal with the serialization format.

---

Let me know if you get stuck on any of this.

**Comment by gsmoon97** [comment]
Hi @kaoutar55, thank you for the detailed feedback. Please find the provenance audit results below.

## Phase 1 provenance audit — complete

Traced a reference **`SimpleMLP`** (1 hidden layer, `128→256→128`, fp16, torch 2.11) through every stage of the Spyre compilation pipeline and recorded which provenance fields exist on the **actual compile-path objects** at each stage. Measurements are taken in-process during one cache-defeated `torch.compile` via read-only hooks on the real objects (FX nodes → `ComputedBuffer` → `OpSpec` → emitted `sdsc_*.json`) — no `torch.export`, no heuristics.

Harness + full write-ups land in `examples/provenance/` via **#2720**. Two docs: `provenance_audit.md` (measurement-only) and `provenance_analysis.md` (interpretation, field semantics, flow diagram, drop locations).

### Acceptance criteria

- [x] Reference model traced through all six stages (model → FX pre/post-grad → Inductor passes → LoopLevelIR → OpSpec → SuperDSC JSON)
- [x] Per field: presence, data structure/format, how populated (copied/derived/generated), and whether it survives to the next stage
- [x] Exact code locations where metadata is dropped
- [x] Pipeline diagram of metadata flow (in `provenance_analysis.md`)
- [x] Stage × field table (below)

### Stage × Field matrix

✓ present (all) · ◐ n/N present on some · ✗ present-able here but not carried (a drop, or an empty slot) · – not applicable (not generated yet, or — for FX-meta downstream — carried only indirectly via `origins`).

| Layer | Field | FX Graph (pre-grad) | FX Graph (post-grad) | LoopLevelIR (pre-pass) | LoopLevelIR (post-pass) | OpSpec | SuperDSC |
| --- | --- | --- | --- | --- | --- | --- | --- |
| FX | `stack_trace` | ✓ | ◐ 3/7 | – | – | – | – |
| FX | `nn_module_stack` | ◐ 2/3 | ◐ 2/7 | – | – | – | – |
| FX | `source_fn_stack` | ✓ | ◐ 3/7 | – | – | – | – |
| FX | `original_aten` | – | ✓ | – | – | – | – |
| FX | `from_node` | – | ◐ 3/7 | – | – | – | – |
| IR | `origins` | – | – | ✓ | ✓ | ✗ | ✗ |
| IR | `origin_node` | – | – | ✓ | ◐ 5/7 | ✗ | ✗ |
| IR | `traceback` | – | – | ✗ | ✗ | ✗ | ✗ |
| IR | `get_stack_traces` | – | – | ◐ 3/5 | ◐ 3/7 | ✗ | ✗ |

`SimpleMLP` compiles to **2 fused SuperDSC kernels** (6 ops + 1).

### Exact drop locations

1. **AOTAutograd `addmm` decomposition** — `linear` → `permute` + `mm` + `add`; the synthesized `mm`/`add` keep `original_aten` but lose `stack_trace`/`nn_module_stack`/`source_fn_stack`.
2. **`origin_node` nulled in pre-scheduling** — set on all buffers entering `CustomPreSchedulingPasses` (`passes.py`), `null` on the two matmul buffers on exit (`origins` survives).
3. **DROP — OpSpec** — `create_op_spec` (`spyre_kernel.py:487`) reads `current_node.node.origins`, but the `OpSpec` dataclass (`op_spec.py:51`) declares no provenance field, so it is discarded.
4. **DROP — SuperDSC JSON** — `parse_op_spec`/`compile_op_spec` (`codegen/superdsc.py`) serialize only `OpSpec` fields; inspected `sdsc_*.json` carry zero provenance.

### Answer: "if I add a `debug_handle` at OpSpec, what can I derive it from?"

**`origins`** is the carrier — present on 7/7 ops at the IR stages, and every other field hangs off it (`origin_node` ⊂ `origins`; `get_stack_traces` is derived from `origins`; the FX-meta is reachable via `origins[*].meta`). From it you can always derive `original_aten`, and `stack_trace`/`nn_module_stack`/`source_fn_stack`/`from_node` for the anchor ops. Do **not** rely on `origin_node` (fragile) or `from_node` (run-dependent). Equivalently, `get_kernel_metadata(node_schedule)` at `define_kernel` already assembles this per kernel. → feeds **#2575** (debug_handle schema on OpSpec/SuperDSC).
